From 0047ac276dbb5612f6974fdd7fcc34155715fd63 Mon Sep 17 00:00:00 2001
From: Mark Enriquez <enriquezmark36@gmail.com>
Date: Tue, 26 May 2020 17:08:45 +0800
Subject: [PATCH 09/12] libcamera: scx15: Port sprd_easy_hdr

Spreadtrums easy HDR library can be used by setting this flag to true:
TARGET_BOARD_CAMERA_HDR_CAPTURE

Change-Id: I68ba83340bcbf8ad879ac861e4df0d7d2302a138
Signed-off-by: Mark Enriquez <enriquezmark36@gmail.com>
---
 libcamera/scx15/Camera.mk                  | 21 ++++++++++++++++++++-
 libcamera/scx15/sc8830/inc/SprdOEMCamera.h |  4 ++++
 libcamera/scx15/sc8830/src/cmr_arith.c     |  9 ++++++++-
 3 files changed, 32 insertions(+), 2 deletions(-)

diff --git a/libcamera/scx15/Camera.mk b/libcamera/scx15/Camera.mk
index c85c663..311ec77 100644
--- a/libcamera/scx15/Camera.mk
+++ b/libcamera/scx15/Camera.mk
@@ -259,13 +259,21 @@ endif
 LOCAL_MODULE_TAGS := optional
 
 ifeq ($(strip $(sc8830like)),1)
-LOCAL_SHARED_LIBRARIES := libexif libutils libbinder libcamera_client libskia libcutils libsqlite libhardware libmorpho_easy_hdr libcamera_metadata libmemoryheapion
+LOCAL_SHARED_LIBRARIES := libexif libutils libbinder libcamera_client libskia libcutils libsqlite libhardware libcamera_metadata libmemoryheapion
 LOCAL_SHARED_LIBRARIES += \
 	libnativewindow \
 	libgui \
 	liblog
 endif
 
+ifeq ($(strip $(TARGET_BOARD_CAMERA_HDR_CAPTURE)),true)
+LOCAL_CFLAGS += -DCONFIG_SPRD_HDR_LIB
+LOCAL_SHARED_LIBRARIES += libsprd_easy_hdr
+else
+LOCAL_SHARED_LIBRARIES += \
+	libmorpho_easy_hdr
+endif
+
 ifdef CONFIG_CAMERA_ISP
 LOCAL_SHARED_LIBRARIES += libisp
 endif
@@ -293,6 +301,17 @@ LOCAL_PREBUILT_LIBS := arithmetic/sc8830/libmorpho_easy_hdr.so
 LOCAL_MODULE_TAGS := optional
 include $(BUILD_MULTI_PREBUILT)
 
+ifeq ($(strip $(TARGET_BOARD_CAMERA_HDR_CAPTURE)),true)
+	include $(CLEAR_VARS)
+	LOCAL_MODULE := libsprd_easy_hdr
+	LOCAL_MODULE_CLASS := SHARED_LIBRARIES
+	LOCAL_MODULE_TAGS := optional
+	LOCAL_MODULE_SUFFIX := .so
+	LOCAL_SRC_FILES := arithmetic/sc8830/libsprd_easy_hdr.so
+	LOCAL_MODULE_PATH := $(TARGET_COPY_OUT_VENDOR)/lib
+	include $(BUILD_PREBUILT)
+endif
+
 endif
 
 
diff --git a/libcamera/scx15/sc8830/inc/SprdOEMCamera.h b/libcamera/scx15/sc8830/inc/SprdOEMCamera.h
index df14237..322eed0 100644
--- a/libcamera/scx15/sc8830/inc/SprdOEMCamera.h
+++ b/libcamera/scx15/sc8830/inc/SprdOEMCamera.h
@@ -26,7 +26,11 @@ extern "C"
 #include <video/sprd_dma_copy_k.h>
 
 #include "../../arithmetic/sc8830/inc/FaceSolid.h"
+#ifdef CONFIG_SPRD_HDR_LIB
+#include "../../arithmetic/sc8830/inc/HDR_SPRD.h"
+#else
 #include "../../arithmetic/sc8830/inc/HDR2.h"
+#endif
 #include "cmr_v4l2.h"
 #include "isp_app.h"
 #include "sensor_drv_u.h"
diff --git a/libcamera/scx15/sc8830/src/cmr_arith.c b/libcamera/scx15/sc8830/src/cmr_arith.c
index 8edf539..c8195e3 100644
--- a/libcamera/scx15/sc8830/src/cmr_arith.c
+++ b/libcamera/scx15/sc8830/src/cmr_arith.c
@@ -290,7 +290,7 @@ void *arithmetic_fd_thread_proc(void *data)
 		}
 
 		if (message.alloc_flag) {
-			free(message.data);
+			free(message.data);
 		}
 
 		if (1 == fd_exit_flag) {
@@ -494,6 +494,9 @@ int arithmetic_hdr_init(uint32_t pic_width, uint32_t pic_height)
 	} else {
 		s_hdr_cxt->mem_size = size;
 	}
+#ifdef CONFIG_SPRD_HDR_LIB
+	sprd_hdr_pool_init();
+#endif
 	return ret;
 }
 
@@ -523,6 +526,10 @@ int arithmetic_hdr_deinit(void)
 	pthread_mutex_unlock(&s_arith_cxt->hdr_lock);
 	pthread_mutex_destroy(&s_arith_cxt->hdr_lock);
 
+#ifdef CONFIG_SPRD_HDR_LIB
+	sprd_hdr_pool_destroy();
+#endif
+
 	CMR_LOGD("e.");
 	return ret;
 }
-- 
2.27.0

